components/movementsensor/mpu6050/mpu6050.go

```diff
@@ -0,0 +1,291 @@
+	return int(int16(binary.BigEndian.Uint16(data)))
+}
+
+// Given a value, scales it so that the range of int16s becomes the range of +/- maxValue.
+func setScale(value int, maxValue float64) float64 {
+	return float64(value) * maxValue / (1 << 15)
+}
+
+// A helper function to abstract out shared code: takes 6 bytes and gives back AngularVelocity.
+func toAngularVelocity(data []byte) spatialmath.AngularVelocity {
+	gx := toSignedValue(data[0:2])
+	gy := toSignedValue(data[2:4])
+	gz := toSignedValue(data[4:6])
+
+	maxRotation := 250.0 // Degrees per second
+	return spatialmath.AngularVelocity{
+		X: setScale(gx, maxRotation),
+		Y: setScale(gy, maxRotation),
+		Z: setScale(gz, maxRotation),
+	}
+}
+
+// A helper function that takes 6 bytes and gives back linear acceleration.
+func toLinearAcceleration(data []byte) r3.Vector {
+	x := toSignedValue(data[0:2])
+	y := toSignedValue(data[2:4])
+	z := toSignedValue(data[4:6])
+
+	// The scale is +/- 2G's, but our units should be mm/sec/sec.
+	maxAcceleration := 2.0 * 9.81 /* m/sec/sec */ * 1000.0 /* mm/m */
+	return r3.Vector{
+		X: setScale(x, maxAcceleration),
+		Y: setScale(y, maxAcceleration),
+		Z: setScale(z, maxAcceleration),
+	}
+}
+
+func (mpu *mpu6050) AngularVelocity(ctx context.Context, extra map[string]interface{}) (spatialmath.AngularVelocity, error) {
+	mpu.mu.Lock()
+	defer mpu.mu.Unlock()
+
+	// The gyroscope measurements are in registers 67 to 72 (2 bytes for each axis).
+	rawData, err := mpu.readBlock(ctx, 67, 6)
+	if err != nil {
+		return spatialmath.AngularVelocity{}, err
+	}
+	return toAngularVelocity(rawData), nil
+}
+
+func (mpu *mpu6050) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	return r3.Vector{}, movementsensor.ErrMethodUnimplementedLinearVelocity
+}
+
+func (mpu *mpu6050) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	return nil, movementsensor.ErrMethodUnimplementedOrientation
+}
+
+func (mpu *mpu6050) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	return 0, movementsensor.ErrMethodUnimplementedCompassHeading
+}
+
+func (mpu *mpu6050) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	return geo.NewPoint(0, 0), 0, movementsensor.ErrMethodUnimplementedPosition
+}
+
+func (mpu *mpu6050) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (mpu *mpu6050) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	mpu.mu.Lock()
+	defer mpu.mu.Lock()
+
+	readings := make(map[string]interface{})
+	// Instead of this function calling AngularVelocity, we get an atomic snapshot of all the
+	// values together, rather than the angular velocity at one time and everything else at a
+	// slightly different time.
```

Comment:
[major] An altrenate way of doing this is to have a for loop in your new function and store the values in the struct. I prefer that way because I always want to be reading continuously in the background somewhere. But if this updates fine with a low end chip, that's fine.

Reply:
Rand and I discussed: certain algorithms want to poll the sensor as fast as possible, and for those you want to ensure that you're not stuck waiting for the mutex to the I2C bus, and can instead give back a sensor reading as quickly as possible. I'll go make this change...